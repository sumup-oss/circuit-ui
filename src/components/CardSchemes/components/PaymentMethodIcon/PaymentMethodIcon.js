/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { keys } from 'lodash/fp';

import { sizes } from '../../../../styles/constants';
import schemeMap from '../../card-scheme-icons';

const { BYTE, KILO, MEGA, GIGA } = sizes;

const wrapperStyles = ({ theme, size }) => css`
  height: ${theme.iconSizes[size]};
  width: auto;
  max-width: ${theme.spacings.zetta};
`;

const IconWrapper = styled('div')(wrapperStyles);

/**
 * A single card scheme icon
 */
const PaymentMethodIcon = ({ iconId, size, ...rest }) => {
  const IconSvg = schemeMap[iconId];

  if (!IconSvg) {
    return null;
  }

  return (
    <IconWrapper size={size} {...rest}>
      <IconSvg
        css={css`
          width: auto;
          max-width: 100%;
          height: 100%;
          display: inline-block;
          line-height: 0;
        `}
      />
    </IconWrapper>
  );
};

PaymentMethodIcon.BYTE = BYTE;
PaymentMethodIcon.KILO = KILO;
PaymentMethodIcon.MEGA = MEGA;
PaymentMethodIcon.GIGA = GIGA;

PaymentMethodIcon.propTypes = {
  /**
   * The id of the card scheme icon.
   */
  iconId: PropTypes.oneOf(keys(schemeMap)),
  /**
   * The optional sizes of the scheme icon
   */
  size: PropTypes.oneOf([
    PaymentMethodIcon.BYTE,
    PaymentMethodIcon.KILO,
    PaymentMethodIcon.MEGA,
    PaymentMethodIcon.GIGA
  ])
};

PaymentMethodIcon.defaultProps = {
  iconId: '',
  size: PaymentMethodIcon.GIGA
};

/**
 * @component
 */
export default PaymentMethodIcon;
