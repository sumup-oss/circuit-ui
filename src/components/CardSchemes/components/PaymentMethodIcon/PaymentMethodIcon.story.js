/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { keys } from 'lodash/fp';
import { select } from '@storybook/addon-knobs';

// eslint-disable-next-line max-len
import schemeMap from '../../../CreditCardDetails/components/scheme-icons/card-scheme-icons';
import PaymentMethodIcon from './PaymentMethodIcon';

const iconSizes = [
  PaymentMethodIcon.BYTE,
  PaymentMethodIcon.KILO,
  PaymentMethodIcon.MEGA,
  PaymentMethodIcon.GIGA
];

export default {
  title: 'Icons/PaymentMethodIcon',
  component: PaymentMethodIcon,
  parameters: {
    jest: ['PaymentMethodIcon']
  }
};

export const base = () => (
  <PaymentMethodIcon
    size={select('Schemes Size', iconSizes, iconSizes[PaymentMethodIcon.GIGA])}
    iconId={select('Schemes Icon', keys(schemeMap), 'visa')}
  />
);
