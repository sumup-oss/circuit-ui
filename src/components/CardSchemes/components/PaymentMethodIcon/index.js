import PaymentMethodIcon from './PaymentMethodIcon';

export default PaymentMethodIcon;
