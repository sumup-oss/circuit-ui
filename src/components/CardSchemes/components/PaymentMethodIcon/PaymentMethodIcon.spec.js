/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PaymentMethodIcon from './PaymentMethodIcon';

describe('PaymentMethodIcon', () => {
  const iconId = 'visa';
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(
      <PaymentMethodIcon size={PaymentMethodIcon.KILO} iconId={iconId} />
    );
    expect(actual).toMatchSnapshot();
  });

  /**
   * Logic tests
   */
  it('should render the specified icon', () => {
    const { getByTestId } = render(
      <PaymentMethodIcon
        size={PaymentMethodIcon.BYTE}
        iconId={iconId}
        data-testid="payment-method-icon"
      />
    );

    const iconEl = getByTestId('payment-method-icon');
    expect(iconEl).toBeVisible();
    expect(iconEl.firstChild).toHaveTextContent(`${iconId}.svg`);
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <PaymentMethodIcon size={PaymentMethodIcon.GIGA} iconId={iconId} />
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
