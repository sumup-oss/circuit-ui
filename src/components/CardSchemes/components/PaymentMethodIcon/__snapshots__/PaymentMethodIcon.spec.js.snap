// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentMethodIcon should render with default styles 1`] = `
.circuit-1 {
  height: 16px;
  width: auto;
  max-width: 56px;
}

.circuit-0 {
  width: auto;
  max-width: 100%;
  height: 100%;
  display: inline-block;
  line-height: 0;
}

<div
  class="circuit-1 circuit-2"
>
  <div
    aria-label="icon visa"
    class="circuit-0"
    role="img"
  >
    visa.svg
  </div>
</div>
`;
